Source/msg.cpp

```diff
@@ -1104,92 +1122,128 @@ DWORD OnRangedAttackPlayer(const TCmd *pCmd, Player &player)
 
 DWORD OnSpellMonster(const TCmd *pCmd, Player &player)
 {
-	auto *p = (TCmdParam4 *)pCmd;
+	const auto &message = *reinterpret_cast<const TCmdParam4 *>(pCmd);
 
-	if (gbBufferMsgs != 1 && currlevel == player.plrlevel) {
-		auto spell = static_cast<spell_id>(p->wParam2);
-		auto spellType = static_cast<spell_type>(p->wParam3);
-		if (currlevel != 0 || spelldata[spell].sTownSpell) {
-			ClrPlrPath(player);
-			player.destAction = ACTION_SPELLMON;
-			player.destParam1 = p->wParam1;
-			player.destParam2 = p->wParam4;
-			player._pSpell = spell;
-			player._pSplType = spellType;
-			player._pSplFrom = 0;
-		} else {
-			PlayerMessageFormat(fmt::format(_("{:s} has cast an illegal spell."), player._pName).c_str());
-		}
-	}
+	if (gbBufferMsgs == 1)
+		return sizeof(message);
+	if (currlevel != player.plrlevel)
+		return sizeof(message);
+	if (message.wParam1 >= MAXMONSTERS)
+		return sizeof(message);
+	if (message.wParam2 > SPL_LAST)
+		return sizeof(message);
+	if (message.wParam3 > RSPLTYPE_INVALID)
+		return sizeof(message);
+
+	auto spell = static_cast<spell_id>(message.wParam2);
+	if (currlevel == 0 && !spelldata[spell].sTownSpell) {
+		LogError(_("{:s} has cast an illegal spell."), player._pName);
+		return sizeof(message);
+	}
+
+	ClrPlrPath(player);
+	player.destAction = ACTION_SPELLMON;
+	player.destParam1 = message.wParam1;
+	player.destParam2 = message.wParam4;

```

Comment:
Should we validate spellLevel ?
Cause `ScaleSpellEffect` could overflow.
(I didn't check if a negative damage has a negative effect. Just wanted to mention it.)

Reply:
I feel that for now it's only necessary to validate engine integration and not game rules. Trying to prevent chats this way is probably largely ineffective, complicated, and risks introducing game play bugs.

Other then the gameplay issue of healing monsters I don't think there is much consequence of overflowing the spell level.

Reply:
For example (not the case in the game): the range of nova could depend on spell level and to check the range there is a array int RangeInTilesForSpellLevel[20]; This could overflow.
I'm not aware of any case there something like this could happen. Just wanted to mention that I didn't checked all places where a too great value could cause a (security) problem.
So I'm okay with it. 🙂

Reply:
Gave it a second look and it all looks good :)